Allow using type names in place of class names

TamasSzigeti opened the pull request:
## Introduce bundle config to map keys to class names

See https://github.com/dunglas/doctrine-json-odm/issues/63

The static type mapper might seem to be weird but this way we don't need to inject anything extra into the serializer and we do not break BC, especially given the doc encourages overriding the service definition.

Comments welcome, let me know what you think, I'll add tests if it can go forward

TamasSzigeti commented:
@dunglas Please have a look

TamasSzigeti commented:
bump @dunglas What do you think of this approach?

bpolaszek commented:
I personally often change class namespaces to better organize code, leading to thorny migrations, so that change would be very welcome!

I +1 👍

dunglas commented on src/Bundle/DependencyInjection/Configuration.php:
```diff
@@ -0,0 +1,40 @@
+                    ->scalarPrototype()
+                        ->cannotBeEmpty()
+                        ->validate()
+                            ->ifTrue(static function ($v): bool {
```
```suggestion
                            ->ifTrue(static function (string $v): bool {
```

dunglas commented on src/Bundle/DependencyInjection/DunglasDoctrineJsonOdmExtension.php:
```diff
@@ -50,5 +50,8 @@ public function load(array $configs, ContainerBuilder $container): void
         }
+
+        $config = $this->processConfiguration(new Configuration(), $configs);
+        $container->setParameter('dunglas_doctrine_json_odm.type_map', $config['types']);
```
Couldn't we create a `TypeMapper` service instead? By using an optional service (null if this option isn't used), we'll probably also improve the performance by preventing some function calls on the hot path.

dunglas commented on src/SerializerTrait.php:
```diff
@@ -27,7 +27,8 @@ public function normalize($data, $format = null, array $context = [])
 
         if (\is_object($data)) {
-            $typeData = [self::KEY_TYPE => \get_class($data)];
+            $typeName = TypeMapper::getTypeByClass(\get_class($data));
```
Same here, couldn't we inject the type mapper as a service instead of calling a static method?

dunglas commented on src/TypeMapper.php:
```diff
@@ -0,0 +1,43 @@
+     */
+    public static function getTypeByClass(string $class): string
+    {
+        $type = array_search($class, self::$map);
```
To avoid the call to `array_search` (which is O(n)), we could store the inverted map in another private property.

TamasSzigeti commented:
@dunglas Addressed change reqs, added tests and notes in readme.

dunglas commented on src/Bundle/DependencyInjection/Configuration.php:
```diff
@@ -0,0 +1,40 @@
+
+        $treeBuilder->getRootNode()
+            ->children()
+                ->arrayNode('types')
```
type_mapping ?

dunglas commented on src/Bundle/DependencyInjection/DunglasDoctrineJsonOdmExtension.php:
```diff
@@ -50,5 +52,15 @@ public function load(array $configs, ContainerBuilder $container): void
+
+        $typeMapConfig = $config['types'];
+        if ($typeMapConfig) {
+            $container->setDefinition(
```
Usually we prefer defining the service in the XML file and removing the definition with `removeDefinition()` in the extension. This allow to have service auto-completion in PHPStorm and the like.

dunglas commented on src/TypeMapper.php:
```diff
@@ -0,0 +1,55 @@
+/**
+ * Allows using string constants in place of class names.
+ */
+class TypeMapper implements TypeMapperInterface
```
```suggestion
final class TypeMapper implements TypeMapperInterface
```

dunglas reviewed (changes requested):
Awesome! I left some minor comments but overall it looks good to me.

TamasSzigeti commented:
@dunglas Also fixed some unrelated issues to make CI green

amenophis commented:
@dunglas any chance to see this merge soon ?
Can I help with something ?

amenophis commented:
@TamasSzigeti Do you want to make the modification proposed by @dunglas ?
If you don't have time, i can do it !

Let me know 🙂

TamasSzigeti commented:
Thanks for the review, Kévin!

@amenophis Thanks for the offer, go ahead if you wish, otherwise I can take care of it this weekend.

amenophis commented:
Hi @TamasSzigeti,

Seems you did not had time to push update.
Could you give me permissions to push on your fork and will try to fix last comments ?

Thanks

TamasSzigeti commented:
@dunglas Comments addressed

@amenophis Thanks, just did it (:

Guite commented:
Really nice feature :+1:

dunglas commented on README.md:
````diff
@@ -154,6 +154,65 @@ $foo = $entityManager->find(Foo::class, $foo->getId());
+        bar: App\SomethingElse\Bar
+```
+
+With this, `Foo` objects will be serialized as
````
```suggestion
With this, `Foo` objects will be serialized as:

```

dunglas commented on README.md:
```diff
@@ -154,6 +154,65 @@ $foo = $entityManager->find(Foo::class, $foo->getId());
+
+#### Using Symfony
+
+In order to use type aliases, add the bundle configuration, e.g. in `config/packages/doctrine_json_odm.yaml`:
```
```suggestion
In order to use type aliases, add the bundle configuration, e.g. in `config/packages/doctrine_json_odm.yaml`:

```

dunglas reviewed (changes requested):
LGTM

TamasSzigeti commented:
@dunglas Comments addressed plus I fixed the workflow again:
- Switched to using mysql from container as somehow setting legacy auth started failing.
- Enable tests on php 8.2
- doctrine/annotations has to become a direct dependency for latest versions

Might be worth dropping unsupported php and symfony versions in another PR

dunglas commented:
Thank you very much for your hard work @TamasSzigeti!!
